src/storage.ts

```diff
@@ -331,6 +339,49 @@ export async function getScript(
     )
 }
 
+const SHA_PART = "sha"
+function binarySha(hex: string) {
+    const s = createHash("sha256")
+    s.update(Buffer.from(hex, "hex"))
+    const sha = s.digest("hex")
+    return sha
+}
+
+export async function resolveScriptBodyFromSha(
+    sha: string
+): Promise<ScriptBody> {
+    if (!sha) return undefined
+
+    // check sha format, roundtrip throug buffer parser
+    sha = Buffer.from(sha, "hex").toString("hex")
+
+    const res = await scriptVersionShaTable.getEntity(SHA_PART, sha)
+    if (!res) return undefined
+
+    const scriptId = res.scriptId as string
+    const scriptVersion = res.scriptVersion as number
+    return await getScriptBody(scriptId, scriptVersion)
+}
+
+async function upsertScriptVersionShaSnapshot(
+    scriptId: string,
+    scriptVersion: number,
+    sha: string
+) {
+    // check sha format, roundtrip throug buffer parser
+    sha = Buffer.from(sha, "hex").toString("hex")
+    // TODO: avoid sha collisions
+    await scriptVersionShaTable.upsertEntity(
+        {
+            partitionKey: SHA_PART,
```

Comment:
I think it should use the same partition (user) key as the script

Reply:
... but I don't know the script is for the lookup

Reply:
messages from device have partitionId, so around `// TODO PARSE MESSAGE` you have `m.partitionId`